Race condition in UPnP code

Doesn't show up when running the integration tests on the build server because there is no UPnP there. :disappointed: 

```
WARNING: DATA RACE
Read by goroutine 47:
  main.(*upnpSvc).tryIGDs()
      /Users/jb/src/github.com/syncthing/syncthing/cmd/syncthing/upnpsvc.go:71 +0xb5
  main.(*upnpSvc).Serve()
      /Users/jb/src/github.com/syncthing/syncthing/cmd/syncthing/upnpsvc.go:41 +0x16e
  github.com/thejerf/suture.func·007()
      /Users/jb/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/thejerf/suture/suture.go:516 +0x14d

Previous write by goroutine 48:
  github.com/syncthing/syncthing/internal/upnp.func·001()
      /Users/jb/src/github.com/syncthing/syncthing/internal/upnp/upnp.go:121 +0x664

Goroutine 47 (running) created at:
  github.com/thejerf/suture.(*Supervisor).runService()
      /Users/jb/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/thejerf/suture/suture.go:519 +0x173
  github.com/thejerf/suture.(*Supervisor).Serve()
      /Users/jb/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/thejerf/suture/suture.go:426 +0xdce

Goroutine 48 (running) created at:
  github.com/syncthing/syncthing/internal/upnp.Discover()
      /Users/jb/src/github.com/syncthing/syncthing/internal/upnp/upnp.go:129 +0x43e
  main.(*upnpSvc).Serve()
      /Users/jb/src/github.com/syncthing/syncthing/cmd/syncthing/upnpsvc.go:38 +0x108
  github.com/thejerf/suture.func·007()
      /Users/jb/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/thejerf/suture/suture.go:516 +0x14d
==================
--- FAIL: TestRestartSenderAndReceiverDuringTransfer (30.78s)
```

--- a/internal/upnp/upnp.go
+++ b/internal/upnp/upnp.go
@@ -103,32 +103,8 @@ func Discover(timeout time.Duration) []IGD {
 
 	resultChan := make(chan IGD)
 
-	// Aggregator
-	go func() {
-	next:
-		for result := range resultChan {
-			for _, existingResult := range results {
-				if existingResult.uuid == result.uuid {
-					if debug {
-						l.Debugf("Skipping duplicate result %s with services:", result.uuid)
-						for _, svc := range result.services {
-							l.Debugf("* [%s] %s", svc.serviceID, svc.serviceURL)
-						}
-					}
-					goto next
-				}
-			}
-			results = append(results, result)
-			if debug {
-				l.Debugf("UPnP discovery result %s with services:", result.uuid)
-				for _, svc := range result.services {
-					l.Debugf("* [%s] %s", svc.serviceID, svc.serviceURL)
-				}
-			}
-		}
-	}()
-
 	wg := sync.NewWaitGroup()
+
 	for _, intf := range interfaces {
 		// Interface flags seem to always be 0 on Windows
 		if runtime.GOOS != "windows" && (intf.Flags&net.FlagUp == 0 || intf.Flags&net.FlagMulticast == 0) {
@@ -144,8 +120,33 @@ func Discover(timeout time.Duration) []IGD {
 		}
 	}
 
-	wg.Wait()
-	close(resultChan)
+	go func() {
+		wg.Wait()
+		close(resultChan)
+	}()
+
+nextResult:
+	for result := range resultChan {
+		for _, existingResult := range results {
+			if existingResult.uuid == result.uuid {
+				if debug {
+					l.Debugf("Skipping duplicate result %s with services:", result.uuid)
+					for _, svc := range result.services {
+						l.Debugf("* [%s] %s", svc.serviceID, svc.serviceURL)
+					}
+				}
+				continue nextResult
+			}
+		}
+
+		results = append(results, result)
+		if debug {
+			l.Debugf("UPnP discovery result %s with services:", result.uuid)
+			for _, svc := range result.services {
+				l.Debugf("* [%s] %s", svc.serviceID, svc.serviceURL)
+			}
+		}
+	}
 
 	return results
 }

